README.md
![raylib-cpp Logo](projects/Doxygen/raylib-cpp_256x256.png)

# raylib-cpp [![Tests](https://github.com/RobLoach/raylib-cpp/workflows/Tests/badge.svg)](https://github.com/RobLoach/raylib-cpp/actions?query=workflow%3ATests+branch%3Amaster) [![License](https://img.shields.io/badge/license-zlib%2Flibpng-blue.svg)](LICENSE)

[raylib-cpp](https://github.com/robloach/raylib-cpp) is a C++ wrapper library for [raylib](https://www.raylib.com), a simple and easy-to-use library to enjoy videogames programming. This C++ header provides object-oriented wrappers around *raylib*'s struct interfaces. *raylib-cpp* is not required to use *raylib* in C++, but the classes do bring using the raylib API more inline with C++'s language pardigm.

## Example

``` cpp
#include "raylib-cpp.hpp"

int main() {
    int screenWidth = 800;
    int screenHeight = 450;

    raylib::Window window(screenWidth, screenHeight, "raylib-cpp - basic window");
    raylib::Texture logo("raylib_logo.png");

    SetTargetFPS(60);

    while (!window.ShouldClose())
    {
        BeginDrawing();

        window.ClearBackground(RAYWHITE);

        DrawText("Congrats! You created your first window!", 190, 200, 20, LIGHTGRAY);

        // Object methods.
        logo.Draw(
            screenWidth / 2 - logo.GetWidth() / 2,
            screenHeight / 2 - logo.GetHeight() / 2);

        EndDrawing();
    }

    // UnloadTexture() and CloseWindow() are called automatically.

    return 0;
}
```

See the [examples folder](examples) for more of the raylib examples that have been ported over to *raylib-cpp*.

## Features

There are a few conventions that raylib-cpp takes on when adopting raylib. See the [raylib-cpp documentation](https://robloach.github.io/raylib-cpp/index.html) for details on the entire API.

### Constructors

Object constructors load raylib objects.

``` cpp
// raylib
Texture2D texture = LoadTexture("texture.png");

// raylib-cpp
raylib::Texture2D texture("texture.png");
```

### Object Methods

When a raylib method has an object as one of its arguments, you can call the method on the object itself.

``` cpp
// raylib
Vector2 position(50, 50);
DrawPixelV(position, PURPLE);

// raylib-cpp
raylib::Vector2 position(50, 50);
position.DrawPixel(PURPLE);
```

### Method Names

If a method's name contains an object's name, it is removed from its name to shorten the method name.

``` cpp
// raylib
DrawTexture(texture, 50, 50, WHITE);

// raylib-cpp
texture.Draw(50, 50, WHITE);
```

### Optional Parameters

Many methods have optional parameters with sane defaults.

``` cpp
// raylib
DrawTexture(texture, 50, 50, WHITE);

// raylib-cpp
texture.Draw(50, 50); // WHITE is provided as the default tint.
```

### Object Destructors

Objects will attempt to unload their respective raylib resources on destruction. This means no need to call Unload or Close methods. This applies to the window, textures, images, sounds, etc.

``` cpp
// raylib
InitWindow(640, 480, "Hello World");
CloseWindow();

// raylib-cpp
raylib::Window window(640, 480, "Hello World");
// window.Close() will be called automatically when the object is destructed.
```

### Property Get/Set

Properties can be assigned through getter and setter methods. You still have access to the internal properties, however.

``` cpp
// raylib
Vector2 position;
position.x = 50;
position.y = 100;

// raylib-cpp
raylib::Vector2 position;
position.SetX(50);
position.SetY(100);

// ... or
position.x = 50;
position.y = 100;
```

### Method Overrides

Many similar raylib method names have different name suffixes based on what arguments they take. With raylib-cpp, these cases use [method overriding](https://en.wikipedia.org/wiki/Method_overriding) to allow using the same method names.

``` cpp
// raylib
Color color = GRAY;
DrawPixel(50, 50, color);
Vector2 position = {50.0f, 50.0f};
DrawPixelV(position, color); // Extra V in method name.

// raylib-cpp
raylib::Color color = raylib::Color::Gray();
color.DrawPixel(50, 50);
Vector2 position(50.0f, 50.0f);
color.DrawPixel(position); // No more V in method name.
position.DrawPixel(color); // Alternatively
```

### Method Chaining

When there's a method that doesn't return anything, it'll instead return the object itself, allowing [method chaining](https://en.wikipedia.org/wiki/Method_chaining).

``` cpp
// raylib
Image cat = ImageLoad("cat.png");
ImageCrop(&cat, (Rectangle){ 100, 10, 280, 380 });
ImageFlipHorizontal(&cat);
ImageResize(&cat, 150, 200);

// raylib-cpp
raylib::Image cat("cat.png");
cat.Crop(100, 10, 280, 380)
   .FlipHorizontal()
   .Resize(150, 200);
```

### Operators

There are operator overrides for objects.

``` cpp
// raylib
Vector2 position = {50, 50};
Vector2 speed = {10, 10};
position.x += speed.x;
position.y += speed.y;

// raylib-cpp
raylib::Vector2 position(50, 50);
raylib::Vector2 speed(10, 10);
position += speed; // Addition assignment operator override.
```

### Vector-Based Returns

When there is a function that would return a pointer-array, there is a wrapper that allows returning a [vector](https://www.cplusplus.com/reference/vector/vector/) of objects instead.

``` cpp
// raylib
int count;
char** files = GetDirectoryFiles(".", &count);
printf("Count: %i\n", count);
for (int i = 0; i < count; i++) {
    printf("File: %s\n", files[i]);
}
ClearDirectoryFiles();

// raylib-cpp
std::vector<std::string> files = raylib::GetDirectoryFiles(".");
std::cout << "Count: " << files.size() << std::endl;
for (auto& file : files) {
    std::cout << "File: " << file << std::endl;
}
```

### String Functions

Many of the raylib functions have `std::string`-related overrides to allow calling them directly with `std::string`s to save having to use the `.c_str()` method.

``` cpp
// raylib
const char* url = "https://raylib.com";
OpenURL(url);

// raylib-cpp
std::string url = "https://raylib.com";
raylib::OpenURL(url);
```

### Exceptions

When loading an asset fails, raylib will log a warning, but provide no other feedback for you to act upon. raylib-cpp will throw a [`RaylibException`](include/RaylibException.hpp) [runtime exception](https://en.cppreference.com/w/cpp/error/runtime_error), allowing to provide a safe method to catch failed loads.

``` cpp
// raylib
Texture texture = LoadTexture("FileNotFound.png");
if (texture.width == 0) {
    TraceLog(LOG_ERROR, "Texture failed to load!");
}

// raylib-cpp
try {
    raylib::Texture texture("FileNotFound.png");
}
catch (raylib::RaylibException& error) {
    error.TraceLog(LOG_ERROR);
}
```

### RayMath

The [raymath](https://github.com/raysan5/raylib/blob/master/src/raymath.h) methods are included.

``` cpp
// raylib
Vector2 direction = {50, 50};
Vector2 newDirection = Vector2Rotate(direction, 30);

// raylib-cpp
raylib::Vector2 direction(50, 50);
raylib::Vector2 newDirection = direction.Rotate(30);
```

## Getting Started

*raylib-cpp* is a header-only library. This means in order to use it, you must link your project to [raylib](https://www.raylib.com/), and then include [`raylib-cpp.hpp`](raylib-cpp/include/raylib-cpp.hpp).

1. Set up a *raylib* project using the [build and install instructions](https://github.com/raysan5/raylib#build-and-installation)
2. Ensure C++ files are compiled with C++
3. Download *raylib-cpp*
4. Include [`include/raylib-cpp.hpp`](include/raylib-cpp.hpp)
    ``` cpp
    #include "path/to/raylib-cpp.hpp"
    ```

### Starter Projects

The [projects directory](projects) includes some starter templates...

- [CMake template](projects/CMake)
- [Make template](projects/Make)
- [VSCode template](projects/VSCode)

If there's a project template you would like to see added, feel free to [make an issue](https://github.com/RobLoach/raylib-cpp/issues) and we can add it in.

### Applications

- [Ian Pan's Raylib Games](https://github.com/ianpan870102/raylib-practices)

## Development

The following are some tools in order to build and contribute to *raylib-cpp*...

### Compiling

Since *raylib-cpp* is a header only library, the build process is the same as raylib's, except you will use C++ to compile instead of C. The following are some specific instructions on local development.

#### Desktop

*raylib-cpp* uses [CMake](https://cmake.org) as a primary target for development. To build it, and run the tests or examples, use...

``` bash
git clone https://github.com/RobLoach/raylib-cpp.git
cd raylib-cpp
mkdir build
cd build
cmake ..
make
make test
./examples/core_basic_window
```

#### Web

Use [emscripten](https://emscripten.org/) to build and test [core_basic_window_web.cpp](examples/core/core_basic_window_web.cpp).

```
mkdir build
cd build
emcmake cmake .. -DPLATFORM=Web -DCMAKE_BUILD_TYPE=Release -DCMAKE_EXE_LINKER_FLAGS="-s USE_GLFW=3"
emmake make
```

See [`core_basic_window_web.html`](examples/core/resources/core_basic_window_web.html) for an example HTML canvas you can you.

### Documentation

To build the document with [Doxygen](http://www.doxygen.nl/), use...

```
doxygen projects/Doxygen/Doxyfile
```

### Coding Standards

This uses cpplint to adopt coding standards.

```
cpplint --recursive include
```

### Defines

- `RAYLIB_CPP_NO_MATH` - When set, will skip adding the `raymath.h` integrations

## License

*raylib-cpp* is licensed under an unmodified zlib/libpng license, which is an OSI-certified, BSD-like license that allows static linking with closed source software. Check [LICENSE](LICENSE) for further details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [4.1.0] - 2022-02-25
### Added
- Web-based example at [core_basic_window_web.cpp](examples/core/core_basic_window_web.cpp)
- An [Exception](README.md#exceptions) is now thrown when assets fail to load during object construction (applies to Music, Texture, Image, Sound, etc)

### Fixed
- TextSubtext() address bounding by @RobLoach in https://github.com/RobLoach/raylib-cpp/pull/159

### Changed
- Unit tests now use [raylib-assert](https://github.com/RobLoach/raylib-assert)

## [4.0.4] - 2022-01-09
### Fixed
- Renamed CMake project name to `raylib_cpp`

## [4.0.3] - 2022-01-09
### Changed
- CMake now used for raylib dependency, rather than git submodules
- Standardized functions across:
  - Color
  - Material
  - Mesh
  - Model
  - ModelAnimation
  - Physics
  - Ray
  - Rectangle
  - Texture
  - Window
- Renamed `package.json` to `clib.json` to follow [clib](https://github.com/clibs/clib) schema

## [4.0.2] - 2021-12-24
### Added
- `Image::IsLoaded()`
- `Wave::IsLoaded()`
- `Sound::IsLoaded()`
- `Music::IsLoaded()`
- Shader example
- Function overrides for `std::string` support
    - `raylib::DrawText()`
    - `raylib::DrawTextEx()`
    - `raylib::DrawTextPro()`
    - `raylib::TextSubtext()`
    - `raylib::TextReplace()`
    - `raylib::TextInsert()`
    - `raylib::TextAppend()`
    - `raylib::TextFindIndex()`
    - `raylib::TextToUpper()`
    - `raylib::TextToLower()`
    - `raylib::TextToPascal()`
    - `raylib::TextToInteger()`

### Fixed
- Shader constructor
- Vector2 muliplication operator behavior ([#155](https://github.com/RobLoach/raylib-cpp/issues/155))

### Changed
- Merged `Text.hpp` into `Functions.hpp`

## [4.0.1] - 2021-11-20
### Changes
- Fix compiler warning

## [4.0.0] - 2021-11-04
### Changes
- Update to [raylib 4.0.0](https://github.com/raysan5/raylib/releases/tag/4.0.0)
- For tests, replaced catch.hpp with `assert()` calls
- Added assignment operators for objects ([#142](https://github.com/RobLoach/raylib-cpp/pull/142) by [@marciejewiczow](https://github.com/maciejewiczow))
- Replaced `NULL` with `nullptr`
- Fixed memory leak in `ModelAnimation::Load` method

## [3.7.0] - 2021-04-23
### Added
- `Window::ClearBackground()` ([#127](https://github.com/RobLoach/raylib-cpp/pull/127) by [ufrshubham](https://github.com/ufrshubham))
### Changes
- Updated to raylib 3.7.0

## [3.5.0] - 2021-03-24
### Added
- VSCode project template (by [KnockerPulsar](https://github.com/KnockerPulsar))
### Changes
- Updated documentation

## [3.5.0-beta3] - 2021-02-18
### Added
- Added a C++ version of raylib's loading thread example (by [@pkeir]([https://github.com/pkeir))
- Updated documentation
### Fixed
- Made the global wrapped functions static to avoid redeclaration

## [v3.5.0-beta2] - 2021-01-18
### Fixed
- `Font.charsPadding` loading
### Removed
- `DroppedFiles` class replaced with `std::vector<std::string> raylib::GetDroppedFiles()`

## [v3.5.0-beta1] - 2021-01-02
### Added
- `Model.Draw()` and `Model.DrawWires()`
- `models_first_person_maze.cpp` example
- String override functions to allow using `std::string` directly instead of `const char*`
- `std::vector<std::string>` wrappers with `raylib::GetDirectoryFiles()` and `raylib::GetDroppedFiles()`
- `raylib::Color::RayWhite()` static functions to build `Color` objects
- `Rectangle.GetPosition()` and `Rectangle.SetPosition()`
- `Rectangle.GetSize()` and `Rectangle.SetSize()`

### Fixed
- Most objects are now passed by reference
- `Mouse::SetX()` and `Mouse::SetY()` setting incorrect values
- Error protection when unloading images, materials, models and meshes

### Changed
- `Mouse` functions are now `static`. Use `Mouse::SetX()` instead of using `Mouse mouse`.
- `Camera*::BeginMode()` and `Camera*::EndMode()` no longer have 2D/3D in the name

## [v3.5.0-alpha1] - 2020-12-26
### Changed
- Update to raylib 3.5.0

### Added
- Documentation

### Removed
- Static inline Color variables (like `raylib::Color::RayWhite`). Instead, use `::RAYWHITE`.


src/main.c
#include <stdio.h>
#include "libretro.h"
#include "raylib.h"
#include "rRetro.h"

int main(int argc, char* argv[]) {
	if (argc < 3) {
		printf("usage: %s <core> <game>\n", argv[0]);
		return 0;
	}

	core_load(argv[1]);
	core_load_game(argv[2]);

	int screenWidth = 800;
    int screenHeight = 450;

    InitWindow(screenWidth, screenHeight, "raylib [core] example - basic window");

    if (!IsWindowReady()) {
    	printf("Failed to initialize window.");
    	return 0;
    }
    
    //SetTargetFPS(60);
    //--------------------------------------------------------------------------------------

    // Main game loop
    while (!WindowShouldClose())    // Detect window close button or ESC key
    {
        // Update
        //----------------------------------------------------------------------------------
        // TODO: Update your variables here
        //----------------------------------------------------------------------------------

        // Draw
        //----------------------------------------------------------------------------------
        BeginDrawing();

            ClearBackground(RAYWHITE);

            DrawText("Congrats! You created your first window!", 190, 200, 20, LIGHTGRAY);

        EndDrawing();
        //----------------------------------------------------------------------------------
    }

	CloseWindow();
	core_unload();
    return 0;
}